addons/mail/models/mail_mail.py

```diff
@@ -425,21 +442,37 @@ def _prepare_outgoing_list(self, recipients_follower_status=None):
                 'partner_id': partner,
             })
 
-        # prepare attachments: remove attachments if user send the link with the
-        # access_token.
         attachments = self.attachment_ids
+        # Prepare attachments:
+        # Remove attachments if user send the link with the access_token.
+        if body:
+            link_ids = {int(link) for link in re.findall(r'/web/(?:content|image)/([0-9]+)', body)}
+            if link_ids:
+                attachments = attachments - self.env['ir.attachment'].browse(list(link_ids))
+        # Turn remaining attachments into links if they are too heavy and
+        # theirs ownerships are business models (i.e. something != mail.message,
+        # otherwise they will be deleted along with the mail message leading to a 404)
+        if business_record_attachments := attachments.filtered(
+                lambda a: a.res_model and a.res_id and a.res_model != 'mail.message'):
+            estimated_email_size = self._estimated_email_size(headers, body,
+                                                              [a.file_size for a in attachments.sudo()])
+            max_email_size_in_bytes = ((mail_server and mail_server.max_email_size) or
+                                       self.env['ir.mail_server']._get_default_max_email_size()) * 1024 * 1024
```

Comment:
Maybe we could define "_get_max_email_size" on ir.mail_server doing both and then use it in a one-liner here, like
```suggestion
            max_email_size_in_bytes = mail_server or self.env['ir.mail_server'])._get_max_email_size() * 1024 * 1024

   ... (on mail server model)
   def _get_max_email_size(self):
       if self.max_email_size:
           return self.max_email_size
       return int(self.env['ir.config_parameter'].sudo().get_param('mail.outgoing_email_size', '5'))
```